File: agent/java/com/squeeze/agent/net/SqueezeTracker.java
package com.squeeze.agent.net;

import com.squeeze.agent.SqueezeReporter;

import org.json.JSONArray;
import org.json.JSONObject;

import java.io.ByteArrayOutputStream;
import java.util.List;
import java.util.Map;

/**
 * Accumulates one HTTP transaction and emits it as a JSON line to the host.
 * Bodies are captured up to a cap. Thread-safe enough for a single connection's
 * lifecycle (request then response).
 */
public final class SqueezeTracker {
    private static final int BODY_CAP = 256 * 1024;

    private final long id = SqueezeReporter.INSTANCE.nextId();
    private final String url;
    private volatile String method = "GET";
    private final long startedAtMs = System.currentTimeMillis();
    private volatile long responseAtMs = 0;
    private volatile boolean reported = false;

    private Map<String, List<String>> requestHeaders;
    private final ByteArrayOutputStream requestBody = new ByteArrayOutputStream();
    private int statusCode = 0;
    private Map<String, List<String>> responseHeaders;
    private final ByteArrayOutputStream responseBody = new ByteArrayOutputStream();
    private String error;
    private String[] callStack;

    public SqueezeTracker(String url) {
        this.url = url;
        this.callStack = captureStack();
    }

    public void setMethod(String m) { if (m != null) method = m; }
    public void setRequestHeaders(Map<String, List<String>> h) { requestHeaders = h; }

    public void appendRequestBody(byte[] b, int off, int len) {
        synchronized (requestBody) {
            int room = BODY_CAP - requestBody.size();
            if (room > 0) requestBody.write(b, off, Math.min(len, room));
        }
    }

    public void onResponse(int code, Map<String, List<String>> headers) {
        if (responseAtMs == 0) responseAtMs = System.currentTimeMillis();
        statusCode = code;
        responseHeaders = headers;
    }

    public void appendResponseBody(int b) {
        synchronized (responseBody) { if (responseBody.size() < BODY_CAP) responseBody.write(b); }
    }

    public void appendResponseBody(byte[] b, int off, int len) {
        synchronized (responseBody) {
            int room = BODY_CAP - responseBody.size();
            if (room > 0) responseBody.write(b, off, Math.min(len, room));
        }
    }

    public void setError(String e) { error = e; }

    public synchronized void report() {
        if (reported) return;
        reported = true;
        try {
            JSONObject o = new JSONObject();
            o.put("type", "txn");
            o.put("id", String.valueOf(id));
            o.put("method", method);
            o.put("url", url);
            o.put("startedAt", startedAtMs / 1000.0);
            if (responseAtMs > 0) o.put("responseAt", responseAtMs / 1000.0);
            o.put("finishedAt", System.currentTimeMillis() / 1000.0);
            o.put("status", statusCode);
            if (error != null) o.put("error", error);
            o.put("requestHeaders", headersToJson(requestHeaders));
            o.put("responseHeaders", headersToJson(responseHeaders));
            o.put("requestBody", bodyToString(requestBody));
            o.put("responseBody", bodyToString(responseBody));
            o.put("requestSize", requestBody.size());
            o.put("responseSize", responseBody.size());
            if (callStack != null) {
                JSONArray cs = new JSONArray();
                for (String s : callStack) cs.put(s);
                o.put("callStack", cs);
            }
            SqueezeReporter.INSTANCE.emit(o.toString());
        } catch (Throwable t) {
            // never let reporting break the app
        }
    }

    private static JSONObject headersToJson(Map<String, List<String>> headers) {
        JSONObject obj = new JSONObject();
        if (headers == null) return obj;
        try {
            for (Map.Entry<String, List<String>> e : headers.entrySet()) {
                String k = e.getKey() == null ? "" : e.getKey();
                List<String> v = e.getValue();
                obj.put(k, v == null ? "" : String.join(", ", v));
            }
        } catch (Throwable ignored) {}
        return obj;
    }

    private static String bodyToString(ByteArrayOutputStream body) {
        byte[] data;
        synchronized (body) { data = body.toByteArray(); }
        if (data.length == 0) return "";
        try {
            return new String(data, "UTF-8");
        } catch (Exception e) {
            return "<" + data.length + " bytes>";
        }
    }

    private static String[] captureStack() {
        StackTraceElement[] st = Thread.currentThread().getStackTrace();
        int start = Math.min(st.length, 4);  // skip getStackTrace + tracker frames
        java.util.ArrayList<String> out = new java.util.ArrayList<>();
        for (int i = start; i < st.length && out.size() < 24; i++) {
            String c = st[i].getClassName();
            if (c.startsWith("com.squeeze.agent")) continue;
            out.add(st[i].toString());
        }
        return out.toArray(new String[0]);
    }
}

File: agent/kotlin/com/squeeze/capture/SqueezeTracker.kt
package com.squeeze.capture

import org.json.JSONArray
import org.json.JSONObject
import java.io.ByteArrayOutputStream

/** Accumulates one HTTP transaction and emits it as a JSON line to the host. */
class SqueezeTracker(private val url: String) {
    private val id = SqueezeReporter.nextId()
    @Volatile private var method = "GET"
    /** Which HTTP stack produced this transaction ("okhttp3", "okhttp2", "urlconnection").
     *  Capture metadata, not policy: the desktop uses it to tell the user whether a request is
     *  even *divertible* (only okhttp3 is). It cannot be derived from [callStack], which
     *  deliberately strips okhttp/okio frames to show app code. */
    @Volatile private var httpStack: String? = null
    private val startedAtMs = System.currentTimeMillis()
    @Volatile private var responseAtMs = 0L
    @Volatile private var reported = false

    private var requestHeaders: Map<String, List<String>>? = null
    private val requestBody = ByteArrayOutputStream()
    private var statusCode = 0
    private var responseHeaders: Map<String, List<String>>? = null
    private val responseBody = ByteArrayOutputStream()
    private var error: String? = null
    private val callStack: List<String> = captureStack()

    fun setMethod(m: String?) { if (m != null) method = m }
    fun setHttpStack(s: String?) { if (s != null) httpStack = s }
    fun setRequestHeaders(h: Map<String, List<String>>?) { requestHeaders = h }

    fun appendRequestBody(b: ByteArray, off: Int, len: Int) {
        synchronized(requestBody) {
            val room = BODY_CAP - requestBody.size()
            if (room > 0) requestBody.write(b, off, minOf(len, room))
        }
    }

    fun onResponse(code: Int, headers: Map<String, List<String>>?) {
        if (responseAtMs == 0L) responseAtMs = System.currentTimeMillis()
        statusCode = code
        responseHeaders = headers
    }

    fun appendResponseByte(b: Int) {
        synchronized(responseBody) { if (responseBody.size() < BODY_CAP) responseBody.write(b) }
    }

    fun appendResponseBody(b: ByteArray, off: Int, len: Int) {
        synchronized(responseBody) {
            val room = BODY_CAP - responseBody.size()
            if (room > 0) responseBody.write(b, off, minOf(len, room))
        }
    }

    fun setError(e: String?) { error = e }

    /**
     * Permanently suppresses this transaction.
     *
     * Used for exchanges that are Jaca talking to itself rather than traffic the app made — a
     * retry-direct bounce from the desktop, whose real request is re-sent immediately after and
     * reported on its own tracker. Marking it reported (rather than adding a second flag) means
     * a later [report] from the body tee is a no-op too.
     */
    @Synchronized
    fun cancel() { reported = true }

    @Synchronized
    fun report() {
        if (reported) return
        reported = true
        try {
            val o = JSONObject()
            o.put("type", "txn")
            o.put("id", id.toString())
            o.put("method", method)
            httpStack?.let { o.put("httpStack", it) }
            o.put("url", url)
            o.put("startedAt", startedAtMs / 1000.0)
            if (responseAtMs > 0) o.put("responseAt", responseAtMs / 1000.0)
            o.put("finishedAt", System.currentTimeMillis() / 1000.0)
            o.put("status", statusCode)
            error?.let { o.put("error", it) }
            o.put("requestHeaders", headersToJson(requestHeaders))
            o.put("responseHeaders", headersToJson(responseHeaders))
            o.put("requestBody", bodyToString(requestBody))
            o.put("responseBody", bodyToString(responseBody))
            o.put("requestSize", requestBody.size())
            o.put("responseSize", responseBody.size())
            val cs = JSONArray()
            for (s in callStack) cs.put(s)
            o.put("callStack", cs)
            SqueezeReporter.emit(o.toString())
        } catch (t: Throwable) {
            // never let reporting break the app
        }
    }

    private fun headersToJson(headers: Map<String, List<String>>?): JSONObject {
        val obj = JSONObject()
        if (headers == null) return obj
        try {
            for ((k, v) in headers) obj.put(k ?: "", v?.joinToString(", ") ?: "")
        } catch (t: Throwable) {}
        return obj
    }

    private fun bodyToString(body: ByteArrayOutputStream): String {
        val data = synchronized(body) { body.toByteArray() }
        if (data.isEmpty()) return ""
        return try { String(data, Charsets.UTF_8) } catch (e: Exception) { "<${data.size} bytes>" }
    }

    /** App-centric call stack: drop our agent, okhttp/okio, reflection and proxy frames
     *  so the stack starts at the app code that issued the request (like AOSP slices the
     *  okhttp frames off). */
    private fun captureStack(): List<String> {
        val st = Thread.currentThread().stackTrace
        val out = ArrayList<String>()
        for (e in st) {
            if (isInfraFrame(e.className)) continue
            out.add(e.toString())
            if (out.size >= 24) break
        }
        return out
    }

    private fun isInfraFrame(c: String): Boolean =
        c.startsWith("com.squeeze") ||
        c.startsWith("okhttp3.") || c.startsWith("okhttp.") || c.startsWith("com.squareup.okhttp") ||
        c.startsWith("okio.") ||
        c.startsWith("java.lang.reflect.") || c.startsWith("jdk.internal.") ||
        c.startsWith("dalvik.system.") || c.contains("\$Proxy") ||
        c == "java.lang.Thread"

    companion object { private const val BODY_CAP = 1024 * 1024 }
}
